go/nats/client.go

```diff
@@ -295,6 +302,21 @@ func (r *streamReader) Close() (err error) {
 			}
 		}
 	}()
+
+	for _, closeFunc := range r.nestCloseFuncs {
```

Comment:
I don't think that tightly coupling the child "unsubscribe" with the parent is correct here, their lifetimes are independent and closing the root stream could prematurely close the nested child subscription this way.
Each stream should be closed independently of each other.

Reply:
agree. the challenge is that `indexedStreamReader` is being passed around as `wrpc.IndexReader` and not `wrpc.IndexReadCloser`, so `Close()` on `indexedStreamReader` is never called.
`indexedStreamReader` also acts as a stream/buffer where the number of expected messages in the subscription is unknown, otherwise we could do `subscription.AutoUnsubscribe(N_messages)`. However, we'd still need a way to cleanup subscriptions if the caller doesn't read from it.

my thinking was that the parent stream ( `streamReader` ) is the one bound to the `close` callback during wrpc `Handle`, so closing it should release all resources associated with it. ex:

```
resp, stop, err := incoming_handler.Handle(reqCtx, client, req)
defer stop() <------- this one
```

we can refactor it so `streamReader` keeps track of borrowed `indexStreamReader`s instead of `nestCloseFuncs`, but don't think we have another spot to safely Unsubscribe

Reply:
We do actually assert for `io.Closer`, which seems to be the idiomatic Go approach for this sort of stuff.
I think the fix here is to implement `io.Closer` for `indexedStreamReader`, which it is currently missing.

On a side note, the `streamReader`/`indexedStreamReader` distinction is a really unfortunate early prototype artifact, ideally this should either just be a single struct or one should embed the other (and the associated `io.Closer` implementation)